// testcafe/lib/browser/provider/built-in/chrome/index.js
'use strict';

exports.__esModule = true;

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _osFamily = require('os-family');

var _osFamily2 = _interopRequireDefault(_osFamily);

var _url = require('url');

var _runtimeInfo = require('./runtime-info');

var _runtimeInfo2 = _interopRequireDefault(_runtimeInfo);

var _config = require('./config');

var _config2 = _interopRequireDefault(_config);

var _localChrome = require('./local-chrome');

var _cdp = require('./cdp');

var cdp = _interopRequireWildcard(_cdp);

var _clientFunctions = require('../../utils/client-functions');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
    openedBrowsers: {},

    isMultiBrowser: false,

    openBrowser: function openBrowser(browserId, pageUrl, configString) {
        var _this = this;

        return (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee() {
            var runtimeInfo, browserName;
            return _regenerator2.default.wrap(function _callee$(_context) {
                while (1) {
                    switch (_context.prev = _context.next) {
                        case 0:
                            _context.next = 2;
                            return (0, _runtimeInfo2.default)((0, _url.parse)(pageUrl).hostname, configString);

                        case 2:
                            runtimeInfo = _context.sent;
                            browserName = _this.providerName.replace(':', '');


                            runtimeInfo.browserId = browserId;
                            runtimeInfo.browserName = browserName;

                            _context.next = 8;
                            return (0, _localChrome.start)(pageUrl, runtimeInfo);

                        case 8:
                            _context.next = 10;
                            return _this.waitForConnectionReady(browserId);

                        case 10:
                            _context.next = 12;
                            return _this.runInitScript(browserId, _clientFunctions.GET_WINDOW_DIMENSIONS_INFO_SCRIPT);

                        case 12:
                            runtimeInfo.viewportSize = _context.sent;

                            if (!(runtimeInfo.config.headless || runtimeInfo.config.emulation)) {
                                _context.next = 16;
                                break;
                            }

                            _context.next = 16;
                            return cdp.createClient(runtimeInfo);

                        case 16:

                            _this.openedBrowsers[browserId] = runtimeInfo;

                        case 17:
                        case 'end':
                            return _context.stop();
                    }
                }
            }, _callee, _this);
        }))();
    },
    closeBrowser: function closeBrowser(browserId) {
        var _this2 = this;

        return (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2() {
            var runtimeInfo;
            return _regenerator2.default.wrap(function _callee2$(_context2) {
                while (1) {
                    switch (_context2.prev = _context2.next) {
                        case 0:
                            runtimeInfo = _this2.openedBrowsers[browserId];

                            if (!cdp.isHeadlessTab(runtimeInfo)) {
                                _context2.next = 6;
                                break;
                            }

                            _context2.next = 4;
                            return cdp.closeTab(runtimeInfo);

                        case 4:
                            _context2.next = 8;
                            break;

                        case 6:
                            _context2.next = 8;
                            return _this2.closeLocalBrowser(browserId);

                        case 8:
                            if (!(_osFamily2.default.mac || runtimeInfo.config.headless)) {
                                _context2.next = 11;
                                break;
                            }

                            _context2.next = 11;
                            return (0, _localChrome.stop)(runtimeInfo);

                        case 11:

                            delete _this2.openedBrowsers[browserId];

                        case 12:
                        case 'end':
                            return _context2.stop();
                    }
                }
            }, _callee2, _this2);
        }))();
    },
    isLocalBrowser: function isLocalBrowser(browserId, configString) {
        var _this3 = this;

        return (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee3() {
            var config;
            return _regenerator2.default.wrap(function _callee3$(_context3) {
                while (1) {
                    switch (_context3.prev = _context3.next) {
                        case 0:
                            config = _this3.openedBrowsers[browserId] ? _this3.openedBrowsers[browserId].config : (0, _config2.default)(configString);
                            return _context3.abrupt('return', !config.headless);

                        case 2:
                        case 'end':
                            return _context3.stop();
                    }
                }
            }, _callee3, _this3);
        }))();
    },
    takeScreenshot: function takeScreenshot(browserId, path) {
        var _this4 = this;

        return (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee4() {
            var runtimeInfo;
            return _regenerator2.default.wrap(function _callee4$(_context4) {
                while (1) {
                    switch (_context4.prev = _context4.next) {
                        case 0:
                            runtimeInfo = _this4.openedBrowsers[browserId];
                            _context4.next = 3;
                            return cdp.takeScreenshot(path, runtimeInfo);

                        case 3:
                        case 'end':
                            return _context4.stop();
                    }
                }
            }, _callee4, _this4);
        }))();
    },
    resizeWindow: function resizeWindow(browserId, width, height, currentWidth, currentHeight) {
        var _this5 = this;

        return (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee5() {
            var runtimeInfo;
            return _regenerator2.default.wrap(function _callee5$(_context5) {
                while (1) {
                    switch (_context5.prev = _context5.next) {
                        case 0:
                            runtimeInfo = _this5.openedBrowsers[browserId];


                            runtimeInfo.viewportSize.width = currentWidth;
                            runtimeInfo.viewportSize.height = currentHeight;

                            _context5.next = 5;
                            return cdp.resizeWindow({ width: width, height: height }, runtimeInfo);

                        case 5:
                        case 'end':
                            return _context5.stop();
                    }
                }
            }, _callee5, _this5);
        }))();
    },
    hasCustomActionForBrowser: function hasCustomActionForBrowser(browserId) {
        var _this6 = this;

        return (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee6() {
            var _openedBrowsers$brows, config, client;

            return _regenerator2.default.wrap(function _callee6$(_context6) {
                while (1) {
                    switch (_context6.prev = _context6.next) {
                        case 0:
                            _openedBrowsers$brows = _this6.openedBrowsers[browserId], config = _openedBrowsers$brows.config, client = _openedBrowsers$brows.client;
                            return _context6.abrupt('return', {
                                hasCloseBrowser: true,
                                hasResizeWindow: !!client && (config.emulation || config.headless),
                                hasTakeScreenshot: !!client,
                                hasCanResizeWindowToDimensions: false,
                                hasMaximizeWindow: false
                            });

                        case 2:
                        case 'end':
                            return _context6.stop();
                    }
                }
            }, _callee6, _this6);
        }))();
    }
};
module.exports = exports['default'];
